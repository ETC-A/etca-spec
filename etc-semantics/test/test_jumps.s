;-mnaked-reg
;0
; Ported from ../../customasm/test_jumps.asm

test_jmp:
    jmp  test_zero_on
.fail:
    jmp  .fail

test_zero_on:
    mov  r0, 0
    test r0, -1

    jnz  .fail
    jz   test_zero_off
.fail:
    jmp .fail ; program counter will indicate failure

test_zero_off:
    mov  r0, 1
    test r0, -1

    jz   .fail
    jnz  test_negative_on
.fail:
    jmp  .fail

test_negative_on:
    mov  r0, -1
    test r0, -1

    jnn  .fail
    jn   test_negative_off
.fail:
    jmp  .fail

test_negative_off:
    mov  r0, 0
    test r0, -1

    jn   .fail
    jnn  test_carry_on
.fail:
    jmp  .fail

test_carry_on:
    mov  r0, -1
    add  r0, 15

    jnc  .fail
    jc   test_carry_off
.fail:
    jmp  .fail

test_carry_off:
    mov  r0, 15
    add  r0, 15
    jc   .fail
    jnc  test_borrow_on
.fail:
    jmp  .fail

test_borrow_on:
    mov  r0, 0
    cmp  r0, 1
    jnc  .fail
    jc   test_borrow_off
.fail:
    jmp  .fail

test_borrow_off:
    mov  r0, 1
    cmp  r0, 1
    jc   .fail
    jnc  test_overflow_on
.fail:
    jmp  .fail

test_overflow_on:
    mov  r0, 0x7fff
    add  r0, 1
    jnv  .fail
    jv   test_overflow_off
.fail:
    jmp  .fail

test_overflow_off:
    mov  r0, 0x7fff
    sub  r0, 1
    jv   .fail
    jnv  test_equal
.fail:
    jmp  .fail

test_equal:
    cmp  r0, r0
    jne  .fail
    jb   .fail
    ja   .fail
    jl   .fail
    jg   .fail

    je   .s1
    jmp  .fail
.s1:
    jbe  .s2
    jmp  .fail
.s2:
    jae  .s3
    jmp  .fail
.s3:
    jle  .s4
    jmp  .fail
.s4:
    jge  test_not_equal
.fail:
    jmp  .fail

test_not_equal:
    mov  r0, 0
    cmp  r0, 1

    je   .fail
    jne  .s1
    jmp  .fail
.s1:
    jb   .s2
    ja   .s2
    jmp  .fail
.s2:
    jl   test_ucomp_1
    jg   test_ucomp_1
.fail:
    jmp  .fail

test_ucomp_1:
    mov  r0, 10
    cmp  r0, 5
    jb   .fail
    jbe  .fail
    ja   .s1
    jmp  .fail
.s1:
    jae  test_ucomp_2
.fail:
    jmp  .fail

test_ucomp_2:
    mov  r0, 5
    cmp  r0, 10
    ja   .fail
    jae  .fail
    jb   .s
    jmp  .fail
.s:
    jbe  test_ucomp_3
.fail:
    jmp  .fail

test_ucomp_3:
    mov  r0, -10
    cmp  r0, 5
    jb   .fail
    jbe  .fail
    ja   .s
    jmp  .fail
.s:
    jae  test_ucomp_4
.fail:
    jmp  .fail

test_ucomp_4:
    mov  r0, 5
    cmp  r0, -10
    ja   .fail
    jae  .fail
    jb   .s
    jmp  .fail
.s:
    jbe  test_scomp_1
.fail:
    jmp  .fail

test_scomp_1:
    mov  r0, 10
    cmp  r0, 5
    jl   .fail
    jle  .fail
    jg   .s
    jmp  .fail
.s:
    jge  test_scomp_2
.fail:
    jmp  .fail

test_scomp_2:
    mov  r0, 5
    cmp  r0, 10
    jg   .fail
    jge  .fail
    jl   .s
    jmp  .fail
.s:
    jle  test_scomp_3
.fail:
    jmp  .fail

test_scomp_3:
    mov  r0, 5
    cmp  r0, -10
    jl   .fail
    jle  .fail
    jg   .s
    jmp  .fail
.s:
    jge  test_scomp_4
.fail:
    jmp  .fail

test_scomp_4:
    mov  r0, -10
    cmp  r0, 5
    jg   .fail
    jge  .fail
    jl   .s
    jmp  .fail
.s:
    jle  done
.fail:
    jmp  .fail

done:
    mov  r7,1
.hlt:
    jmp  .hlt
