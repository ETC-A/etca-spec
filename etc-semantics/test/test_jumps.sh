set -u
./scripts/test.sh "$1" "$2" "$3" "test/test_jumps.bin" 0
