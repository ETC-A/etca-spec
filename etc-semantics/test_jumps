b"\x8e\x06\x5f\x03\x8e\x00\x59\x00\x57\x1f\x91\xf8\x80\x06\x5f\x03\x8e\x00\x59\x01\x57\x1f\x90\xec\x81\x06\x5f\x03\x8e\x00\x59\x1f\x57\x1f\x93\xe0\x82\x06\x5f\x03\x8e\x00\x59\x00\x57\x1f\x92\xd4\x83\x06\x5f\x03\x8e\x00\x59\x0f\x50\x0f\x94\xc8\x85\x06\x5f\x03\x8e\x00\x59\x1f\x50\x0f\x95\xbc\x84\x06\x5f\x03\x8e\x00\x59\x01\x53\x01\x94\xb0\x85\x06\x5f\x03\x8e\x00\x59\x00\x53\x01\x95\xa4\x84\x06\x5f\x03\x8e\x00\x59\x00\x5c\x1f\x5c\x1f\x5c\x1f\x51\x01\x96\x92\x87\x06\x5f\x03\x8e\x00\x59\x00\x5c\x1f\x5c\x1f\x5c\x1f\x50\x01\x97\x80\x86\x06\x5f\x03\x8e\x00\x13\x00\x91\x76\x94\x74\x99\x72\x9a\x70\x9d\x6e\x80\x06\x5f\x03\x8e\x00\x88\x06\x5f\x03\x8e\x00\x85\x06\x5f\x03\x8e\x00\x8c\x06\x5f\x03\x8e\x00\x8b\x06\x5f\x03\x8e\x00\x59\x00\x53\x01\x90\x4a\x81\x06\x5f\x03\x8e\x00\x84\x08\x89\x06\x5f\x03\x8e\x00\x8a\x08\x8d\x06\x5f\x03\x8e\x00\x59\x0a\x53\x05\x94\x2e\x98\x2c\x89\x06\x5f\x03\x8e\x00\x85\x06\x5f\x03\x8e\x00\x59\x05\x53\x0a\x99\x1a\x95\x18\x84\x06\x5f\x03\x8e\x00\x88\x0a\x5f\x03\x8e\x00\x5f\x03\x8e\x00\x59\x16\x53\x05\x94\xf8\x98\xf6\x89\x06\x5f\x03\x8e\x00\x85\x06\x5f\x03\x8e\x00\x59\x05\x53\x16\x99\xe4\x95\xe2\x84\x06\x5f\x03\x8e\x00\x88\x06\x5f\x03\x8e\x00\x59\x0a\x53\x05\x9a\xd0\x9c\xce\x8d\x06\x5f\x03\x8e\x00\x8b\x06\x5f\x03\x8e\x00\x59\x05\x53\x0a\x9d\xbc\x9b\xba\x8a\x06\x5f\x03\x8e\x00\x8c\x06\x5f\x03\x8e\x00\x59\x05\x53\x16\x9a\xa8\x9c\xa6\x8d\x06\x5f\x03\x8e\x00\x8b\x06\x5f\x03\x8e\x00\x59\x16\x53\x05\x9d\x94\x9b\x92\x8a\x06\x5f\x03\x8e\x00\x8c\x06\x5f\x03\x8e\x00\x8e\x00"