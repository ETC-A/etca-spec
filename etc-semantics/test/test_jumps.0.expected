<etc>
  <k>
    #halt "Halted successfully!" ~> .
  </k>
  <pc> 33068 </pc>
  <registers>
    0 |-> -10
    1 |-> 0
    2 |-> 0
    3 |-> 0
    4 |-> 0
    5 |-> 0
    6 |-> 0
    7 |-> 1
  </registers>
  <memory>
    128 |-> b"\x8e\x04\x8e\x00Y\x00W\x1f\x81\x04\x80\x04\x8e\x00Y\x01W\x1f\x80\x04\x81\x04\x8e\x00Y\x1fW\x1f\x83\x04\x82\x04\x8e\x00Y\x00W\x1f\x82\x04\x83\x04\x8e\x00Y\x1fP\x0f\x85\x04\x84\x04\x8e\x00Y\x0fP\x0f\x84\x04\x85\x04\x8e\x00Y\x00S\x01\x85\x04\x84\x04\x8e\x00Y\x01S\x01\x84\x04\x85\x04\x8e\x00Y\x00\\\x1f\\\x1f\\\x1fP\x01\x87\x04\x86\x04\x8e\x00Y\x00\\\x1f\\\x1f\\\x1fQ\x01\x86\x04\x87\x04\x8e\x00\x13\x00\x81\x1c\x84\x1a\x89\x18\x8a\x16\x8d\x14\x80\x04\x8e\x10\x88\x04\x8e\f\x85\x04\x8e\x08\x8c\x04\x8e\x04\x8b\x04\x8e\x00Y\x00S\x01\x80\x10\x81\x04\x8e\f\x84\x06\x89\x04\x8e\x06\x8a\x06\x8d\x04\x8e\x00Y\nS\x05\x84\n\x88\x08\x89\x04\x8e\x04\x85\x04\x8e\x00Y\x05S\n\x89\n\x85\x08\x84\x04\x8e\x04\x88\x04\x8e\x00Y\x16S\x05\x84\n\x88\x08\x89\x04\x8e\x04\x85\x04\x8e\x00Y\x05S\x16\x89\n\x85\x08\x84\x04\x8e\x04\x88\x04\x8e\x00Y\nS\x05\x8a\n\x8c\x08\x8d\x04\x8e\x04\x8b\x04\x8e\x00Y\x05S\n\x8d\n"
    129 |-> b"\x8b\x08\x8a\x04\x8e\x04\x8c\x04\x8e\x00Y\x05S\x16\x8a\n\x8c\x08\x8d\x04\x8e\x04\x8b\x04\x8e\x00Y\x16S\x05\x8d\n\x8b\x08\x8a\x04\x8e\x04\x8c\x04\x8e\x00Y\xe1\x8e\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
  </memory>
  <flags>
    <carry>    false </carry>
    <zero>     false </zero>
    <negative> true </negative>
    <overflow> false </overflow>
  </flags>
  <machine-details>
    <cpuid1> 0 </cpuid1>
    <cpuid2> 0 </cpuid2>
    <feat>   0 </feat>
    <reg-width> word </reg-width>
    <reg-count> 8 </reg-count>
    <evil-mode> false </evil-mode>
  </machine-details>
  <in> </in>
  <out> </out>
</etc>
